test/test_auth_endpoint.py

```diff
@@ -1,36 +1,46 @@
+from config import test_config
 import pytest
 import requests
 import base64
 import re
 import logging
 import html
 
-# Pragmatic
-# Search for IRODS_HTTP_API_BASE_URL
-# and IRODS_CLIENT_VERSION
-# TOOD: Consider turning this into a fixture
-AUTH_URL = 'http://127.0.0.1:9000/irods-http-api/0.2.0/authenticate'
+skip_protected_resource = pytest.mark.skipif(
+    test_config['openid_connect']['mode'] == 'protected_resource', reason='Not supported in "protected_resource" mode.')
```

Comment:
What happens if the config options aren't defined?

Reply:
Can you be more specific? Do you mean the entire `test_config`? The `mode` option? Yes to all of the above?

Reply:
If `openid_connect` and/or `mode` is not defined, this will result in a lookup error.

The test harness should handle missing options gracefully. Either by exiting with a good log message or falling back to a reasonable default.

For this specific case, logging a good error message and exiting seems like the correct approach.

Thoughts?

Reply:
Would a schema make more sense? Validate early, fail otherwise?

Reply:
I thought the tests didn't have a schema. We're talking about `config.py` not the server, which has its own config. But yes, the idea is to catch and give a stern talking.

Reply:
> we... already have a schema? so at this point, we should just... catch a lookup error and exit angrily?

This is for test's config file (which is just a python file).

> Would a schema make more sense? Validate early, fail otherwise?

That would work. How would you do that with a python file though? Unless you're talking about changing the file to a json file.

Reply:
ah, okay.    carry on with the discussion.

if not present - exit:
 - pro: easier, less confusing to admins, finds errors quickly
 - con: does not allow for easy upgrades if/when we introduce a newly required setting/value

if not present - use default:
 - pro: easier upgrades, more forgiving to admins who didn't get it right
 - con: could be surprising to find things are running with a 'magic' default

i'm leaning to exit early on this one...

Reply:
I'm in the middle on this one.

For some things, it's clear they need to be required. For others, a reasonable default makes sense. For example, the OIDC stuff is required for the OIDC tests, but can be ignored for the non-OIDC tests.

Reply:
I pushed up an example schema. Worked when I ran it.

Reply:
I see the call to `validate()` at the bottom of the python file. When is that run?

Perhaps each test suite needs its own schema?

Reply:
If recall correctly, it's run on import.

> Perhaps each test suite needs its own schema?

You mean besides the HTTP API?

Reply:
No. There are two test suites in this repo. The OIDC tests and the non-OIDC tests.

They have different requirements, therefore it may be worth using different schemas. That will make it clear which options are absolutely necessary for each suite.

If we don't like two schemas, then we can say all options are required and include comments in the config.py explaining what each option is for.

Reply:
Think the two schemas approach was too awkward, given that only the 'openid_connect' option is sectioned out for easy disabling. If we want a multiple schema approach, it would be better to section off the 'required' options and the 'non-oidc' test options.

Reply:
Or... we just make everything required. Then the person running the tests has to make sure everything is configured correctly. No need for anything fancy.

We just need to include good documentation for testing.

That keeps things simple and straight forward.

Thoughts?

Reply:
just saying this out loud... if they're required (rather than present/not), then we have to give them a way to 'turn it off'.

Reply:
Very true. The config.py currently favors the non-oidc tests.

I'm just throwing out ideas here. Who knows if they're good or bad :-)

Reply:
We decided to make all the properties required for the moment.